pkg/issuer/acme/setup_test.go

```diff
@@ -0,0 +1,520 @@
+			kfsShouldBeCalled:          true,
+			kfsErr:                     notFoundErr,
+			acmePrivKeySecretCreateErr: someErr,
+			expectedConditions: []cmapi.IssuerCondition{
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorAccountRegistrationFailed),
+					gen.SetIssuerConditionMessage(messageAccountRegistrationFailed+someErr.Error())),
+			},
+			wantsErr: true,
+		},
+		"ACME private key secret does not exist, account key generation is disabled": {
+			issuer: gen.IssuerFrom(baseIssuer,
+				gen.SetIssuerACMEDisableAccountKeyGeneration(true),
+			),
+			kfsShouldBeCalled: true,
+			kfsErr:            notFoundErr,
+			expectedConditions: []cmapi.IssuerCondition{
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorAccountVerificationFailed),
+					gen.SetIssuerConditionMessage(messageAccountVerificationFailed+messageNoSecretKeyGenerationDisabled+notFoundErr.Error())),
+			},
+			wantsErr: true,
+		},
+		"ACME private key secret exists, but contains invalid private key": {
+			issuer:            gen.IssuerFrom(baseIssuer),
+			kfsShouldBeCalled: true,
+			kfsErr:            invalidDataErr,
+			expectedConditions: []cmapi.IssuerCondition{
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorAccountVerificationFailed),
+					gen.SetIssuerConditionMessage(fmt.Sprintf("%s%v", messageInvalidPrivateKey, invalidDataErr))),
+			},
+		},
+		"Checking ACME private key secret fails with an unknown error": {
+			issuer:            gen.IssuerFrom(baseIssuer),
+			kfsShouldBeCalled: true,
+			kfsErr:            someErr,
+			expectedConditions: []cmapi.IssuerCondition{
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorAccountVerificationFailed),
+					gen.SetIssuerConditionMessage(messageAccountVerificationFailed+someErr.Error())),
+			},
+			wantsErr: true,
+		},
+		"ACME account's key is not an RSA key": {
+			issuer: gen.IssuerFrom(baseIssuer,
+				gen.SetIssuerACMEPrivKeyRef(issuerSecretKeyName)),
+			kfsShouldBeCalled: true,
+			kfsKey:            ecdsaPrivKey,
+			expectedConditions: []cmapi.IssuerCondition{
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorAccountVerificationFailed),
+					gen.SetIssuerConditionMessage(fmt.Sprintf(messageTemplateNotRSA, issuerSecretKeyName))),
+			},
+		},
+		"ACME server URL is an invalid URL": {
+			issuer: gen.IssuerFrom(baseIssuer,
+				gen.SetIssuerACMEURL(invalidURL)),
+			kfsShouldBeCalled:          true,
+			kfsKey:                     rsaPrivKey,
+			removeClientShouldBeCalled: true,
+			expectedConditions: []cmapi.IssuerCondition{
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorInvalidURL),
+					gen.SetIssuerConditionMessage(invalidURLMessage)),
+			},
+			expectedEvents: []string{
+				fmt.Sprintf("%s %s %s", corev1.EventTypeWarning, errorInvalidURL, invalidURLMessage)},
+		},
```

Comment:
:+1: This is another case where I think there should be API webhook validation (perhaps there is) and if so, we should just call those API validation functions here in the setup function or in the parent controller. But this new test is great.

Reply:
I made the comment earlier, but it's not clear to me why events are not also generated for earlier failures in the  account key secret checking and creation.

Reply:
>  I think there should be API webhook validation (perhaps there is) and if so, we should just call those API validation functions here in the setup function or in the parent controller

Currently in the webhook we only validate that the field is set [here](https://github.com/jetstack/cert-manager/blob/master/pkg/internal/apis/certmanager/validation/issuer.go#L112).

I guess if we were to validate the URL itself in the webhook, the validation would have to:
1) error if the field is an invalid URL
2) error and show the correct LetsEncrypt v2 URL if the field is LetsEncrypt v1 URL (with or without the slash)

That would require for the webhook to be aware of the LetsEncrypt v1 -> v2 prod/staging URL mapping and have some sort of URL parser.
Is that perhaps making the webhook too smart? We have also talked about not wanting to put too much into the webhook. But maybe this is the right way to go.

Reply:
I think it's the natural place to perform such validation. And easier to unit test there too. But in a future PR.